daffodil-codegen-c/src/main/resources/org/apache/daffodil/codegen/c/files/libruntime/validators.c

```diff
@@ -38,6 +38,32 @@ validate_array_bounds(const char *name, size_t count, size_t minOccurs, size_t m
     }
 }
 
+// Validate element or sequence's dfdl:assert expression returns true
+
+void
+validate_dfdl_assert(bool assertion, const char *expression, const char *failure_type,
+                     ParserOrUnparserState *pu)
+{
+    // The boolean expression that determines if the assertion is true
+    // is evaluated in generated code and the result passed here, as
+    // well as failure_type which determines whether to issue a diagnostic
+    // which is printed later or an error which terminates processing now.
+    if (!assertion && failure_type[0] == 'r')
+    {
+        Diagnostics *diagnostics = get_diagnostics();
+        const Error error = {ERR_DFDL_ASSERT, {.s = expression}};
+
+        add_diagnostic(diagnostics, &error);
+        pu->diagnostics = diagnostics;
+    }
+    else if (!assertion)
+    {
+        static Error error = {ERR_DFDL_ASSERT, {0}};
```

Comment:
Unrelated to this PR, but just out of curiosity, is there a reason you use static `Error`'s for pu error? I guess an alternative would be to make the `struct Error` in `ParserOrUnparserState` be a non-pointer and do something like:
```c
pu->error.code = ERR_DFDL_ASSERT;
pu->error.s = expression;
```
I have no preference either way, just wondering since I feel using static is less common in C. But I also haven't done C development in a while, so maybe I'm forgetting a subtle difference. Or is it just preference?

And after asking, I now see that `pu->error == NULL` means no error, which you can't do if error is a non-pointer. I guess with the non-pointer approach you would need a special code of ERR_NO_ERROR or something, and do `pu->error.code == ERR_NO_ERR`, which is too verbose. That could be wrapped in a function, e.g. `if (is_error(pstate->pu)) return;`, but it's probably not worth the refactor.

So nevermind. I'll leave this in case others have the same question as I did.